File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/palindrome.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *reverse(Node *head) {
    if (head == nullptr || head->next == nullptr) return head;
    Node *newHead = reverse(head->next);
    Node *front = head->next;
    front->next = head;
    head->next = nullptr;
    return newHead;
}

void checkPalindrome(Node *head) {
    if (head == nullptr || head->next == nullptr) {
        traverseLinkedList(head);
        return;
    }
    Node *fast = head;
    Node *slow = head;
    while (fast->next != nullptr && fast->next->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
    }
    Node *newHead = reverse(slow->next);
    Node *first = head;
    Node *second = newHead;
    while (second != nullptr) {
        if (first->data != second->data) {
            reverse(newHead);
            cout << "Not a Palindrome" << endl;
            return;
        }
        first = first->next;
        second = second->next;
    }
    reverse(newHead);
    cout << "Is a Palindrome" << endl;
}

int main() {
    vector<int> arr = {1, 2, 3, 2, 1};
    Node *head = array_to_linkedlist(arr);
    checkPalindrome(head);
    return 0;
}

File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/Palindrome.java
public class Palindrome {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[]{1, 2, 3, 3, 2, 1});
        checkPalindrome(head);
    }

    static Node reverse(Node head) {
        if (head == null || head.next == null) return head;
        Node newHead = reverse(head.next);
        Node front = head.next;
        front.next = head;
        head.next = null;
        return newHead;
    }

    static void checkPalindrome(Node head) {
        if (head == null || head.next == null) {
            LinkedListOperations.traverseLinkedList(head);
            return;
        }
        Node slow = head;
        Node fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        Node newHead = reverse(slow.next);
        Node first = head;
        Node second = newHead;
        while (second != null) {
            if (first.data != second.data) {
                reverse(newHead);
                System.out.println("Not a Palindrome");
                return;
            }
            first = first.next;
            second = second.next;
        }
        reverse(newHead);
        System.out.println("Is a Palindrome");
    }
}
